File: src/lib/api/scripts.ts
import type {
  LogId,
  Script,
  ScriptId,
  LogsQuery,
  ScriptLog,
  UpdateScript,
  CreateScript,
  UpdateOrdering,
} from "$shared/dataV2";

import { invoke } from "@tauri-apps/api/core";
import { createQuery, createMutation } from "@tanstack/svelte-query";

import { queryClient } from "./utils";

const SCRIPTS_KEY = ["scripts"];

function invalidateScriptsList() {
  return queryClient.invalidateQueries({ queryKey: SCRIPTS_KEY });
}

export function getScripts() {
  return invoke<Script[]>("get_scripts");
}

function createScriptKey(id: ScriptId) {
  return ["script", id] as const;
}

export function getScriptById(scriptId: ScriptId) {
  return invoke<Script | null>("get_script_by_id", { scriptId });
}

export async function createScript(create: CreateScript) {
  const script = await invoke<Script>("create_script", { create });

  // Invalidate the specific script query
  const scriptKey = createScriptKey(script.id);
  queryClient.setQueryData(scriptKey, script);

  invalidateScriptsList();

  return script;
}

export async function updateScript(
  update: UpdateScript,
  invalidateList = true,
) {
  const script = await invoke<Script>("update_script", update);

  // Invalidate the specific item query
  const itemKey = createScriptKey(script.id);
  queryClient.setQueryData(itemKey, script);

  if (invalidateList) invalidateScriptsList();

  return script;
}

export async function deleteScript(scriptId: ScriptId, invalidateList = true) {
  await invoke<void>("delete_script", { scriptId });

  const scriptKey = createScriptKey(scriptId);

  // Cancel any queries for the item and clear the current item data
  queryClient.cancelQueries({ queryKey: scriptKey });
  queryClient.setQueryData(scriptKey, undefined);

  if (invalidateList) invalidateScriptsList();
}

export async function bulkDeleteScripts(scriptIds: ScriptId[]) {
  await Promise.all(scriptIds.map((scriptId) => deleteScript(scriptId, false)));
  invalidateScriptsList();
}

function createScriptLogsKey(id: ScriptId, query?: LogsQuery) {
  if (query === undefined) {
    return ["script-logs", id] as const;
  }

  return ["script-logs", id, query] as const;
}

export function getScriptLogs(scriptId: ScriptId, query: LogsQuery) {
  return invoke<ScriptLog[]>("get_script_logs", { scriptId, query });
}

export function invalidateScriptLogs(scriptId: ScriptId, query: LogsQuery) {
  const queryKey = createScriptLogsKey(scriptId, query);
  queryClient.invalidateQueries({ queryKey });
}

export function scriptLogsQuery(scriptId: ScriptId, query: LogsQuery) {
  return createQuery({
    queryKey: createScriptLogsKey(scriptId, query),
    queryFn: () => getScriptLogs(scriptId, query),
  });
}

export async function deleteScriptLogs(scriptId: ScriptId, logIds: LogId[]) {
  await invoke<void>("delete_script_logs", { logIds });

  queryClient.setQueriesData<ScriptLog[]>(
    { queryKey: createScriptLogsKey(scriptId) },
    (data) => {
      if (data === undefined) return undefined;
      return data.filter((log) => !logIds.includes(log.id));
    },
  );

  // Invalid the list of items
  queryClient.invalidateQueries({
    queryKey: createScriptLogsKey(scriptId),
  });
}

export async function updateScriptOrder(update: UpdateOrdering[]) {
  await invoke("update_script_orderings", { update });

  invalidateScriptsList();
}

// -----------------------------------------------------

export function createScriptsQuery() {
  return createQuery({
    queryKey: SCRIPTS_KEY,
    queryFn: getScripts,
  });
}

export function createScriptQuery(id: ScriptId) {
  return createQuery({
    queryKey: createScriptKey(id),
    queryFn: () => getScriptById(id),
  });
}

export function deleteScriptMutation() {
  return createMutation<void, Error, ScriptId>({
    mutationFn: deleteScript,
  });
}

File: src-tauri/src/commands/scripts.rs
//! # Scripts
//!
//! Commands for interacting with scripts from the frontend

use crate::database::entity::{
    script_logs::ScriptLogsModel,
    scripts::{CreateScript, UpdateScript, UpdateScriptOrdering},
    shared::LogsQuery,
    ScriptModel,
};
use anyhow::Context;
use sea_orm::{DatabaseConnection, ModelTrait};
use tauri::State;
use uuid::Uuid;

use super::CmdResult;

/// Get all scripts
#[tauri::command]
pub async fn get_scripts(db: State<'_, DatabaseConnection>) -> CmdResult<Vec<ScriptModel>> {
    let db = db.inner();
    let scripts = ScriptModel::all(db).await?;
    Ok(scripts)
}

/// Get a specific script by ID
#[tauri::command]
pub async fn get_script_by_id(
    script_id: Uuid,
    db: State<'_, DatabaseConnection>,
) -> CmdResult<Option<ScriptModel>> {
    let db = db.inner();
    let script = ScriptModel::get_by_id(db, script_id).await?;
    Ok(script)
}

/// Create a new script
#[tauri::command]
pub async fn create_script(
    create: CreateScript,
    db: State<'_, DatabaseConnection>,
) -> CmdResult<ScriptModel> {
    let db = db.inner();
    let script = ScriptModel::create(db, create).await?;
    Ok(script)
}

/// Update an existing script
#[tauri::command]
pub async fn update_script(
    script_id: Uuid,
    update: UpdateScript,
    db: State<'_, DatabaseConnection>,
) -> CmdResult<ScriptModel> {
    let db = db.inner();
    let script = ScriptModel::get_by_id(db, script_id)
        .await?
        .context("script not found")?;
    let script = script.update(db, update).await?;
    Ok(script)
}

/// Delete a script
#[tauri::command]
pub async fn delete_script(script_id: Uuid, db: State<'_, DatabaseConnection>) -> CmdResult<()> {
    let db = db.inner();
    let script = ScriptModel::get_by_id(db, script_id)
        .await?
        .context("script not found")?;
    script.delete(db).await?;
    Ok(())
}

/// Get logs of a script
#[tauri::command]
pub async fn get_script_logs(
    script_id: Uuid,
    query: LogsQuery,
    db: State<'_, DatabaseConnection>,
) -> CmdResult<Vec<ScriptLogsModel>> {
    let db = db.inner();
    let script = ScriptModel::get_by_id(db, script_id)
        .await?
        .context("script not found")?;
    let logs = script.get_logs(db, query).await?;

    Ok(logs)
}

#[tauri::command]
pub async fn delete_script_logs(
    log_ids: Vec<Uuid>,
    db: State<'_, DatabaseConnection>,
) -> CmdResult<()> {
    let db = db.inner();

    ScriptLogsModel::delete_many(db, &log_ids).await?;

    Ok(())
}

#[tauri::command]
pub async fn update_script_orderings(
    update: Vec<UpdateScriptOrdering>,
    db: State<'_, DatabaseConnection>,
) -> CmdResult<()> {
    let db = db.inner();
    ScriptModel::update_order(db, update).await?;

    Ok(())
}
